src/hex_tarball.erl

```diff
@@ -43,19 +99,42 @@
 %%        inner_checksum => <<178,12,...>>}}
 %% '''
 %% @end
--spec create(metadata(), files()) -> {ok, {tarball(), checksum()}}.
+-spec create(metadata(), files()) -> {ok, tarball()} | {error, errors(), warnings()}.
 create(Metadata, Files) ->
+    case validate(Metadata, Files) of
+        #{files := Files1, errors := [], warnings := Warnings} ->
+             Metadata1 = filter_metadata([unknown, deprecated], Metadata, Warnings),
+            {ok, Res} = create_tarball(Metadata1, Files1),
+            Final = maps:put(warnings, Warnings, Res),
+            {ok, Final};
+        #{errors := Errors, warnings := Warnings} ->
+            {error, #{errors => Errors, warnings => Warnings}}
+    end.
+
+filter_metadata(Types, Meta, Unwanted) ->
+  lists:foldl(fun({K, T}, Acc) ->
+                  case lists:member(T, Types) of
+                    true ->
+                      maps:remove(K, Meta);
+                    false ->
+                      Acc
+                  end
+              end,
+              Meta,
+              Unwanted).
+
+create_tarball(Metadata, Files) ->
     MetadataBinary = encode_metadata(Metadata),
     ContentsTarball = create_memory_tarball(Files),
     ContentsTarballCompressed = gzip(ContentsTarball),
     InnerChecksum = inner_checksum(?VERSION, MetadataBinary, ContentsTarballCompressed),
     InnerChecksumBase16 = encode_base16(InnerChecksum),
 
     OuterFiles = [
-       {"VERSION", ?VERSION},
-       {"CHECKSUM", InnerChecksumBase16},
-       {"metadata.config", MetadataBinary},
-       {"contents.tar.gz", ContentsTarballCompressed}
+        #{name => "VERSION", contents => ?VERSION},
```

Comment:
I changed this to go with the other `file()` map that is now being slung around to be consistent. I need to add a helper to create a new `file()` map that has all keys if we are in agreement.